File: Decktracker/Decktracker/View/Search/LimitedSearchViewController.m
//
//  LimitedSearchViewController.m
//  Decktracker
//
//  Created by Jovit Royeca on 9/6/14.
//  Copyright (c) 2014 Jovito Royeca. All rights reserved.
//

#import "LimitedSearchViewController.h"
#import "AddCardViewController.h"
#import "Card.h"
#import "FileManager.h"
#import "SearchResultsTableViewCell.h"

#import "GAI.h"
#import "GAIDictionaryBuilder.h"
#import "GAIFields.h"

@implementation LimitedSearchViewController

@synthesize deck = _deck;
@synthesize searchBar = _searchBar;
@synthesize tblResults = _tblResults;
@synthesize predicate = _predicate;
@synthesize fetchedResultsController = _fetchedResultsController;

- (NSFetchedResultsController *)fetchedResultsController
{
    if (_fetchedResultsController != nil)
    {
        return _fetchedResultsController;
    }
    
    NSFetchedResultsController *nsfrc = [[Database sharedInstance] search:self.searchBar.text withPredicate:self.predicate];
    
    self.fetchedResultsController = nsfrc;
    _fetchedResultsController.delegate = self;
    return _fetchedResultsController;
}

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        // Custom initialization
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    CGFloat dX = 0;
    CGFloat dY = 0;
    CGFloat dWidth = self.view.frame.size.width;
    CGFloat dHeight = self.view.frame.size.height;
    self.searchBar = [[UISearchBar alloc] init];
    self.searchBar.autoresizingMask = UIViewAutoresizingFlexibleWidth;
    self.searchBar.delegate = self;
    
    self.tblResults = [[UITableView alloc] initWithFrame:CGRectMake(dX, dY, dWidth, dHeight)
                                                   style:UITableViewStylePlain];
    self.tblResults.delegate = self;
    self.tblResults.dataSource = self;
    [self.tblResults registerNib:[UINib nibWithNibName:@"SearchResultsTableViewCell" bundle:nil]
          forCellReuseIdentifier:@"Cell"];
    
    self.navigationItem.titleView = self.searchBar;
    [self.view addSubview:self.tblResults];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

-(void) viewDidAppear:(BOOL)animated
{
    MBProgressHUD *hud = [[MBProgressHUD alloc] initWithView:self.view];
    [self.view addSubview:hud];
    hud.delegate = self;
    [hud showWhileExecuting:@selector(doSearch) onTarget:self withObject:nil animated:NO];
}

#pragma mark - UISearchBarDelegate
- (void)searchBarSearchButtonClicked:(UISearchBar *)bar
{
    if ([self.searchBar canResignFirstResponder])
    {
        [self.searchBar resignFirstResponder];
    }
    
    MBProgressHUD *hud = [[MBProgressHUD alloc] initWithView:self.view];
    [self.view addSubview:hud];
    hud.delegate = self;
    [hud showWhileExecuting:@selector(doSearch) onTarget:self withObject:nil animated:NO];
    
    // send to Google Analytics
    id tracker = [[GAI sharedInstance] defaultTracker];
    [tracker send:[[GAIDictionaryBuilder createEventWithCategory:@"Simple Search"
                                                          action:self.searchBar.text
                                                           label:@"Run"
                                                           value:nil] build]];
}

- (void) doSearch
{
    self.fetchedResultsController = nil;
    
    NSError *error;
    if (![[self fetchedResultsController] performFetch:&error])
    {
        NSLog(@"Unresolved error %@, %@", error, [error userInfo]);
    }
    
    [self.tblResults reloadData];
}

#pragma mark - MBProgressHUDDelegate methods
- (void)hudWasHidden:(MBProgressHUD *)hud
{
	[hud removeFromSuperview];
}

#pragma mark - UITableView
- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return SEARCH_RESULTS_CELL_HEIGHT;
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    NSInteger count = [[self.fetchedResultsController sections] count];
	return count;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    id <NSFetchedResultsSectionInfo> sectionInfo = [self.fetchedResultsController sections][section];
    
	if (self.fetchedResultsController)
    {
        return [NSString stringWithFormat:@"%tu Results", [sectionInfo numberOfObjects]];
    }
    else
    {
        return nil;
    }
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
	id <NSFetchedResultsSectionInfo> sectionInfo = [self.fetchedResultsController sections][section];
    
	return [sectionInfo numberOfObjects];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *CellIdentifier = @"Cell";
    
    SearchResultsTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    if (cell == nil)
    {
        cell = [[SearchResultsTableViewCell alloc] initWithStyle:UITableViewCellStyleDefault
                                                 reuseIdentifier:CellIdentifier];
    }
    
    Card *card = [self.fetchedResultsController objectAtIndexPath:indexPath];
    [cell displayCard:card];
    
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    AddCardViewController *view = [[AddCardViewController alloc] init];
    
    view.arrDecks = [[NSMutableArray alloc] initWithArray:@[self.deck.name]];
    view.arrCollections = [[NSMutableArray alloc] init];
    for (NSString *file in [[FileManager sharedInstance] listFilesAtPath:@"/Collections"
                                                          fromFileSystem:FileSystemLocal])
    {
        [view.arrCollections addObject:[file stringByDeletingLastPathComponent]];
    }
    
    view.card = [self.fetchedResultsController objectAtIndexPath:indexPath];
    view.showCardButtonVisible = YES;
    view.segmentedControlIndex = 0;
    [self.navigationController pushViewController:view animated:YES];
}

#pragma mark - NSFetchedResultsControllerDelegate
- (void)controllerWillChangeContent:(NSFetchedResultsController *)controller
{
    // The fetch controller is about to start sending change notifications, so prepare the table view for updates.
    [self.tblResults beginUpdates];
    NSLog(@"%d %s %s", __LINE__, __PRETTY_FUNCTION__, __FUNCTION__);
}

- (void)controller:(NSFetchedResultsController *)controller
   didChangeObject:(id)anObject
       atIndexPath:(NSIndexPath *)indexPath
     forChangeType:(NSFetchedResultsChangeType)type
      newIndexPath:(NSIndexPath *)newIndexPath
{
    NSLog(@"%d %s %s", __LINE__, __PRETTY_FUNCTION__, __FUNCTION__);
    
    UITableView *tableView = self.tblResults;
    
    switch(type)
    {
        case NSFetchedResultsChangeInsert:
        {
            [tableView insertRowsAtIndexPaths:[NSArray arrayWithObject:newIndexPath]
                             withRowAnimation:UITableViewRowAnimationFade];
            break;
        }
        case NSFetchedResultsChangeDelete:
        {
            [tableView deleteRowsAtIndexPaths:[NSArray arrayWithObject:indexPath]
                             withRowAnimation:UITableViewRowAnimationFade];
            break;
        }
        case NSFetchedResultsChangeUpdate:
        {
            Card *card = [self.fetchedResultsController objectAtIndexPath:indexPath];
            SearchResultsTableViewCell *cell = (SearchResultsTableViewCell*)[tableView cellForRowAtIndexPath:indexPath];
            [cell displayCard:card];
            break;
        }
        case NSFetchedResultsChangeMove:
        {
            [tableView deleteRowsAtIndexPaths:[NSArray arrayWithObject:indexPath]
                             withRowAnimation:UITableViewRowAnimationFade];
            
            [tableView insertRowsAtIndexPaths:[NSArray arrayWithObject:newIndexPath]
                             withRowAnimation:UITableViewRowAnimationFade];
            break;
        }
    }
}

- (void)controller:(NSFetchedResultsController *)controller
  didChangeSection:(id )sectionInfo
           atIndex:(NSUInteger)sectionIndex
     forChangeType:(NSFetchedResultsChangeType)type
{
    NSLog(@"%d %s %s", __LINE__, __PRETTY_FUNCTION__, __FUNCTION__);
    
    UITableView *tableView = self.tblResults;
    
    switch(type)
    {
        case NSFetchedResultsChangeInsert:
        {
            [tableView insertSections:[NSIndexSet indexSetWithIndex:sectionIndex]
                     withRowAnimation:UITableViewRowAnimationFade];
            break;
        }
        case NSFetchedResultsChangeDelete:
        {
            [tableView deleteSections:[NSIndexSet indexSetWithIndex:sectionIndex]
                     withRowAnimation:UITableViewRowAnimationFade];
            break;
        }
        default:
        {
            break;
        }
    }
}

- (void)controllerDidChangeContent:(NSFetchedResultsController *)controller
{
    NSLog(@"%d %s %s", __LINE__, __PRETTY_FUNCTION__, __FUNCTION__);
    
    UITableView *tableView = self.tblResults;
    
    // The fetch controller has sent all current change notifications, so tell the table view to process all updates.
    [tableView endUpdates];
}

@end

File: ios/old/Decktracker/View/Search/LimitedSearchViewController.swift
//
//  LimitedSearchViewController.swift
//  Decktracker
//
//  Created by Jovit Royeca on 6/9/15.
//  Copyright (c) 2015 Jovito Royeca. All rights reserved.
//

import UIKit

class LimitedSearchViewController: UIViewController, UISearchBarDelegate, UITableViewDataSource, UITableViewDelegate, MBProgressHUDDelegate
{
    var deckName:String?
    var searchBar:UISearchBar?
    var tblResults:UITableView?
    var predicate:NSPredicate?
    var sections:Array<[String: [String]]>?
    var sectionIndexTitles:[String]?
    var sectionName:String?
    var viewLoadedOnce = true
    var searchTimer:NSTimer?
    var placeHolderTitle:String?
    
    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
        
        let dX = CGFloat(0)
        let dY = CGFloat(0)
        let dWidth = self.view.frame.size.width
        let dHeight = self.view.frame.size.height
        self.searchBar = UISearchBar()
        self.searchBar!.autoresizingMask = UIViewAutoresizing.FlexibleWidth
        self.searchBar!.delegate = self;
        // Add a Done button in the keyboard
        let barButton = UIBarButtonItem(barButtonSystemItem: UIBarButtonSystemItem.Done, target: self.searchBar, action: "resignFirstResponder")
        let toolbar = UIToolbar(frame: CGRectMake(0, 0, dWidth, 44))
        toolbar.items = [barButton]
        self.searchBar!.inputAccessoryView = toolbar
        self.searchBar!.placeholder = placeHolderTitle;
        
        self.tblResults = UITableView(frame: CGRectMake(dX, dY, dWidth, dHeight), style: UITableViewStyle.Plain)
        self.tblResults!.delegate = self
        self.tblResults!.dataSource = self

        self.navigationItem.titleView = self.searchBar;
        self.view.addSubview(self.tblResults!)
        
        self.sectionName = "sectionNameInitial"
        sections = Array<[String: [String]]>()
        sectionIndexTitles = [String]()
        self.viewLoadedOnce = false
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    func loadData() {
        let hud = MBProgressHUD(view: view)
        view!.addSubview(hud)
        hud.delegate = self
        hud.showWhileExecuting("doSearch", onTarget: self, withObject: nil, animated: true)
    }
    
    func doSearch() {
        var unsortedSections = [String: [String]]()
        sections = Array<[String: [String]]>()
        sectionIndexTitles = [String]()
        
        let cards = Database.sharedInstance().findCards(self.searchBar!.text, withPredicate:self.predicate, withSortDescriptors: [RLMSortDescriptor(property: "name", ascending: true)], withSectionName:sectionName)
        for x in cards {
            let card = x as! DTCard
            let name = card.sectionNameInitial
            let predicate = NSPredicate(format: "%K = %@", sectionName!, name!)
            
            if (name != nil) {
                var cardIds = Array<String>()
                
                for y in cards.objectsWithPredicate(predicate) {
                    let z = y as! DTCard
                    cardIds.append(z.cardId)
                }
                
                let index = name!.startIndex.advancedBy(1)
                var indexTitle = name!.substringToIndex(index)
                
                if name == "Blue" {
                    indexTitle = "U"
                }
                
                unsortedSections.updateValue(cardIds, forKey: name!)
                if !sectionIndexTitles!.contains(indexTitle) {
                    sectionIndexTitles!.append(indexTitle)
                }
            }
        }
        
        for k in unsortedSections.keys.sort(<) {
            let dict = [k: unsortedSections[k]!]
            sections!.append(dict)
        }
        
        self.tblResults!.reloadData()
    }
    
//    MARK: UISearchBarDelegate
    func searchBar(searchBar: UISearchBar, textDidChange searchText: String) {
        if searchTimer != nil && searchTimer!.valid {
            searchTimer!.invalidate()
        }
        searchTimer = NSTimer(timeInterval: 1.0, target: self, selector: "handleSearchBarEndTyping", userInfo: nil, repeats: false)
        NSRunLoop.mainRunLoop().addTimer(searchTimer!, forMode:NSDefaultRunLoopMode)

    }
    
    func handleSearchBarEndTyping() {
        let hud = MBProgressHUD(view: view)
        view!.addSubview(hud)
        hud.delegate = self;
        
        if searchBar!.text!.isEmpty {
            sections = Array<[String: [String]]>()
            sectionIndexTitles = [String]()
            self.tblResults!.reloadData()
            
        } else {
            hud.showWhileExecuting("loadData", onTarget: self, withObject: nil, animated: true)
        }
    }
    
//    MARK: UITableViewDelegate
    func tableView(tableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
        return CGFloat(CARD_SUMMARY_VIEW_CELL_HEIGHT)
    }
    
    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        let dict = sections![indexPath.section]
        let key = dict.keys.first
        var cardIds = dict[key!]
        let cardId = cardIds![indexPath.row]
        
        let view = AddCardViewController()
        view.arrDecks = NSMutableArray(array: [self.deckName!])
        view.cardId = cardId;
        view.createButtonVisible = false
        view.showCardButtonVisible = true
        view.segmentedControlIndex = 0;
        self.navigationController!.pushViewController(view, animated: true)
    }

//    MARK : UITableViewDataSource
    func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return sections!.count
    }
    
    func tableView(tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        let dict = sections![section]
        let key = dict.keys.first
        let cardIds = dict[key!]
        let cardsString = cardIds!.count > 1 ? "cards" : "card"
        return "\(key) (\(cardIds!.count) \(cardsString))"
    }
    
    func sectionIndexTitlesForTableView(tableView: UITableView) -> [String]? {
        return sectionIndexTitles
    }
    
    func tableView(tableView: UITableView, sectionForSectionIndexTitle title: String, atIndex index: Int) -> Int {
        return index
    }
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let dict = sections![section]
        let key = dict.keys.first
        return dict[key!]!.count
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let dict = sections![indexPath.section]
        let key = dict.keys.first
        let cardIds = dict[key!]
        let cardId = cardIds![indexPath.row]
        var cell:UITableViewCell?
        var cardSummaryView:CardSummaryView?
        
        if let x = tableView.dequeueReusableCellWithIdentifier(kCardInfoViewIdentifier) as UITableViewCell! {
            cell = x
            for subView in cell!.contentView.subviews {
                if subView is CardSummaryView {
                    cardSummaryView = subView as? CardSummaryView
                    break
                }
            }
            
        } else {
            cell = UITableViewCell(style: UITableViewCellStyle.Default, reuseIdentifier: kCardInfoViewIdentifier)
            cardSummaryView = NSBundle.mainBundle().loadNibNamed("CardSummaryView", owner: self, options: nil).first as? CardSummaryView
            cardSummaryView!.frame = CGRect(x: 0, y: 0, width: tableView.frame.size.width, height: CGFloat(CARD_SUMMARY_VIEW_CELL_HEIGHT))
            cell!.contentView.addSubview(cardSummaryView!)
        }
        
        cell!.accessoryType = UITableViewCellAccessoryType.None
        cell!.selectionStyle = UITableViewCellSelectionStyle.None
        cardSummaryView!.displayCard(cardId)
        
        return cell!
    }
    
//    MARK: MBProgressHUDDelegate methods
    func hudWasHidden(hud: MBProgressHUD) {
        hud.removeFromSuperview()
    }

}
